Don't use/verify Cinder capabilities if no backend

bogdando opened the pull request:
Cinder volume and backup should have supported storage backends
configured in order to use and verify volumes/snapshots with a
test VM that we start before adoption.

When the volume backend is not supported, do not attach volumes
and do not create snapshots for the test/bfv VMs.

When the backup backend is not supported, do not create volumes
backups.

Neither verify, if those are still present post-adoption.

Depends-on: https://review.rdoproject.org/r/c/rdo-jobs/+/55084

fmount commented on tests/roles/development_environment/files/pre_launch.bash:
```diff
@@ -67,6 +67,8 @@ export FIP=192.168.122.20
 # check connectivity via FIP
 ping -c4 ${FIP}
 
+if [ "$CINDER_BACKEND_CONFIGURED" = "false" ]; then exit 0; fi
```
wondering if it would be better to wrap from here to the end w/ a cinder related function, but really a minor thing wrt the flow of the script.

bogdando replied:
I really don't know, if putting a large code block under extra indentaion would look better. I think I am inspired by Golang here, that avoids such things :)

jistr replied:
I would also prefer what Francesco suggested. Someone may want to add an extra thing at the end and not notice the early exit. To me the early exit is a surprising design. But it's not a blocker, if it passes CI we can merge.

rajathere replied:
I would also prefer the cinder commands added inside a block as well since reading 20+ lines before where we want to add new code seems not usual.
This script has become the main source of validation for all resources before -> after adoption so I won't be surprised if other projects start creating resources below this.
If the indentation is an issue then we can add the following lines in a function and call it based on the CINDER_BACKEND_CONFIGURED value.

bogdando replied:
done

bogdando commented:
oops, this needs more testing
```
TASK [development_environment : pre-launch test VM instance] *******************
fatal: [localhost]: FAILED! => {"msg": "template error while templating string: Could not load \"lowercase\": 'lowercase'. String: {{ shell_header }}\nexport OPENSTACK_COMMAND=\"{{ openstack_command }}\"\nexport CINDER_BACKEND_CONFIGURED=\"{{ cinder_backend_configured | string | lowercase }}\"\n{{ lookup('ansible.builtin.file', prelaunch_test_instance_script) }}\n. Could not load \"lowercase\": 'lowercase'"}
```

jistr commented on tests/roles/development_environment/tasks/main.yaml:
```diff
@@ -1,10 +1,13 @@
   no_log: "{{ use_no_log }}"
   when: prelaunch_test_instance|bool
+  vars:
+    cinder_backend_configured: "{{  cinder_volume_backend == 'ceph' and cinder_backup_backend == 'ceph' }}"
```
Wondering if it would be better to use `cinder_volume_backend == '' or cinder_backup_backend == ''`? If we add testing for another backend like NFS, this wouldn't have to change then. Similarly with the `cinder_verify.yaml`, perhaps we should be checking `cinder_volume_backend != ''` rather than `cinder_volume_backend == 'ceph'` ?

rajathere replied:
is the cinder_backup_backend validation needed? If we don't have cinder_volume configured then we can't create volumes and eventually can't create backups as well.
Also agree that when adding jobs for other backends like NFS or iSCSI this might need change so we should not stick to ceph being the only valid backend available for cinder.

bogdando replied:
I found a better solution, which is adding a var that we can use differently upstream and downstream

rajathere commented on tests/roles/dataplane_adoption/tasks/cinder_verify.yaml:
```diff
@@ -5,7 +5,9 @@
-  when: prelaunch_test_instance|bool
+  when:
+    - prelaunch_test_instance|bool
+    - cinder_volume_backend == 'ceph'
```
cinder_volume_backend != ''

rajathere commented on tests/roles/dataplane_adoption/tasks/cinder_verify.yaml:
```diff
@@ -16,7 +18,9 @@
-  when: prelaunch_test_instance|bool
+  when:
+    - prelaunch_test_instance|bool
+    - cinder_backup_backend == 'ceph'
```
cinder_volume_backend != ''

rajathere commented on docs_dev/assemblies/development_environment.adoc:
```diff
@@ -229,9 +229,12 @@ Create a test VM instance with a test volume attachement:
 OS_CLOUD_IP=192.168.122.100 OS_CLOUD_NAME=standalone \
     bash tests/roles/development_environment/files/pre_launch.bash
 ----
+<1> Use `CINDER_BACKEND_CONFIGURED=false`, if Cinder Volume and Backup services' storage backends have been not configured for the source cloud,
```
Also it's possible to have a volume backend without a backup backend so again we might end up in a situation where volumes can be created but backups cannot.
A lot of our customer environments use volume backends but doesn't use OpenStack provided backup backends and rather use third party solutions.

bogdando replied:
done

rajathere commented on tests/vars.sample.yaml:
```diff
@@ -56,6 +56,19 @@ ironic_adoption: false
+  - fc
+
+supported_backup_backends: #CUSTOMIZE_THIS
+  - ceph
+  - iscsi
+  - nfs
+  - fc
```
a correct list would be

- ceph
- nfs
- s3
- swift

rajathere commented on tests/roles/development_environment/files/pre_launch.bash:
```diff
@@ -17,6 +17,44 @@ function wait_for_status() {
+function create_backup_resources() {
+    # create volume backup
+    if ! ${BASH_ALIASES[openstack]} volume backup show backup; then
+        ${BASH_ALIASES[openstack]} volume backup create --name backup disk
```
${BASH_ALIASES[openstack]} volume backup create --name backup disk --force

but I'm wondering if we really want to create the backup after creating server from volume.
backing up before server create might make this more tedious by adding another function and calling it.
I'm okay with the current workflow just we need the --force flag here.

rajathere commented on tests/roles/dataplane_adoption/tasks/cinder_verify.yaml:
```diff
@@ -4,22 +4,26 @@
 
-- name: Verify that the volumes have correct name, status and size
-  when: prelaunch_test_instance|bool
+- name: Verify that the volumes and snapshots have correct name, status and size
```
hmm, these validations doesn't seem to be running, did we miss something anywhere?
this is the adoption-standalone-to-crc-ceph job

TASK [dataplane_adoption : Verify that the volumes and snapshots have correct name, status and size] ***
skipping: [localhost] => {"changed": false, "false_condition": "cinder_volume_backend in supported_volume_backends", "skip_reason": "Conditional result was False"}

TASK [dataplane_adoption : verify the backup of disk volume] *******************
skipping: [localhost] => {"changed": false, "false_condition": "cinder_backup_backend in supported_backup_backends", "skip_reason": "Conditional result was False"}

bogdando replied:
yea, the ceph job got
```
export CINDER_VOLUME_BACKEND_CONFIGURED=\"false\"
export CINDER_BACKUP_BACKEND_CONFIGURED=\"false\"
```

bogdando replied:
zuul provided [inventory](https://logserver.rdoproject.org/58/658/4551da92f65083ba6899ec1a67c24ff17f0bf107/github-check/adoption-standalone-to-crc-ceph/b29ade9/zuul-info/inventory.yaml) has no supported backends defined from this dependency https://review.rdoproject.org/r/c/rdo-jobs/+/55084/2/zuul.d/_data_plane_adoption.yaml#34

not sure why

bogdando commented:
Note, must merge dowstream PR as well, before merging this one!

bogdando commented:
Not sure why, the tasks wasn't triggered again
```
TASK [dataplane_adoption : Verify that the volumes and snapshots have correct name, status and size] ***
skipping: [localhost] => {"changed": false, "false_condition": "cinder_volume_backend in supported_volume_backends",
"skip_reason": "Conditional result was False"}

TASK [dataplane_adoption : verify the backup of disk volume] *******************
skipping: [localhost] => {"changed": false, "false_condition": "cinder_backup_backend in supported_backup_backends",
"skip_reason": "Conditional result was False"}
```
See https://logserver.rdoproject.org/58/658/b54eeb9032adb1c3b7e10cd678544ea94a4eccc4/github-check/adoption-standalone-to-crc-ceph/57b21a4/controller/data-plane-adoption-tests-repo/data-plane-adoption/tests/logs/test_with_ceph_out_2024-10-17T11:24:33EDT.log

bogdando commented:
recheck

bogdando commented:
recheck

bogdando commented:
recheck after https://github.com/openstack-k8s-operators/install_yamls/pull/937 done

bogdando commented:
The ceph standalone config looks correct https://logserver.rdoproject.org/58/658/b54eeb9032adb1c3b7e10cd678544ea94a4eccc4/github-check/adoption-standalone-to-crc-ceph/bed673c/controller/data-plane-adoption-tests-repo/data_plane_adoption/vars.yaml

So does the deployemnt https://logserver.rdoproject.org/58/658/b54eeb9032adb1c3b7e10cd678544ea94a4eccc4/github-check/adoption-standalone-to-crc-ceph/bed673c/controller/data-plane-adoption-tests-repo/data-plane-adoption/tests/logs/test_with_ceph_out_2024-10-24T09:09:05EDT.log 
```
TASK [dataplane_adoption : Verify that the volumes and snapshots have correct name, status and size] ***
changed: [localhost] => {"attempts": 1, "changed": true, "cmd": "alias openstack=\"oc exec -t openstackclient -- openstack\"

${BASH_ALIASES[openstack]} volume list -c Name -c Status -c Size -f value | grep \"disk in-use 1\"
${BASH_ALIASES[openstack]} volume list -c Name -c Status -c Size -f value | grep \"boot-volume in-use 1\"
${BASH_ALIASES[openstack]} volume snapshot list -c Name -c Status -c Size -f value | grep \"snapshot available 1\"
",
"delta": "0:00:09.402921",
"end": "2024-10-24 10:37:25.909397",
"rc": 0, "start": "2024-10-24 10:37:16.506476",
"stdout": "disk in-use 1
boot-volume in-use 1
snapshot available 1",
"stdout_lines": ["disk in-use 1",
"boot-volume in-use 1",
"snapshot available 1"]}

TASK [dataplane_adoption : verify the backup of disk volume] *******************
changed: [localhost] => {"attempts": 1, "changed": true, "cmd": "alias openstack=\"oc exec -t openstackclient -- openstack\"
```

@fmount @jistr this one is good to go!

rajathere commented:
Thanks @bogdando for the work on this.
The changes looks good and the job results verify the expected behavior.
LGTM.

bogdando commented:
https://review.rdoproject.org/r/c/rdo-jobs/+/55084 was forgotten, needs a fast merge

jistr commented:
Included in https://github.com/openstack-k8s-operators/data-plane-adoption/pull/709
